[JIT] fix dict update

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45929 [JIT] Fix Dict bug in constant hashing
* **#45857 [JIT] fix dict update**

Fix for https://github.com/pytorch/pytorch/issues/45627

Op was calling `insert` instead of `insert_or_assign`, so it wouldn't overwrite an existing key.

Differential Revision: [D24148805](https://our.internmc.facebook.com/intern/diff/D24148805)

Commit:
fix bad api call

[ghstack-poisoned]

gmagogsfm commented on test/jit/test_list_dict.py:
```diff
@@ -1288,6 +1288,14 @@ def update(a, b):
         self.checkScript(update, (self.dict(), self.dict()))
         self.checkScript(update, (self.dict(), self.dict2()))
 
+        def upd() -> Dict[str, int]:
```
Nit, make it into its own testcase for better parallel execution/clearer error message etc.

SplitInfinity replied:
```suggestion
        def update_existing_key() -> Dict[str, int]:
```

SplitInfinity commented on test/jit/test_list_dict.py:
```diff
@@ -1288,6 +1288,14 @@ def update(a, b):
         self.checkScript(update, (self.dict(), self.dict2()))
 
+        def upd() -> Dict[str, int]:
+            a: Dict[str, int] = {}
+            for i in range(3):
+                a.update({'a': i})
```
Do we need several iterations? Would something like this suffice:
```suggestion
            a: Dict[str, int] = {'a': 0}
            a.update({'a': 1})
```

Commit:
Update on "[JIT] fix dict update"


Fix for https://github.com/pytorch/pytorch/issues/45627

Op was calling `insert` instead of `insert_or_assign`, so it wouldn't overwrite an existing key.

[ghstack-poisoned]

Commit:
Update on "[JIT] fix dict update"


Fix for https://github.com/pytorch/pytorch/issues/45627

Op was calling `insert` instead of `insert_or_assign`, so it wouldn't overwrite an existing key.

[ghstack-poisoned]

Commit:
Update on "[JIT] fix dict update"


Fix for https://github.com/pytorch/pytorch/issues/45627

Op was calling `insert` instead of `insert_or_assign`, so it wouldn't overwrite an existing key.

Differential Revision: [D24148805](https://our.internmc.facebook.com/intern/diff/D24148805)

[ghstack-poisoned]
